neo/Network/P2P/Payloads/Cosigner.cs

```diff
@@ -0,0 +1,76 @@
+﻿using Neo.Cryptography.ECC;
+using Neo.IO;
+using Neo.IO.Json;
+using System;
+using System.IO;
+using System.Linq;
+
+namespace Neo.Network.P2P.Payloads
+{
+    public class Cosigner : ISerializable
+    {
+        public UInt160 Account;
+        public WitnessScope Scopes;
```

Comment:
@shargon Should prop `Scopes`'s type be `WitnessScope[]`, as you have named scopes not scope.
And there are cases where multiple scopes is necessary.

Reply:
You can add multiple cosigners

Reply:
@shargon May I ask what's `sender`'s scope? Is it Global? If so, what if I want the sender to have scoped witness, too?

Reply:
`Sender` pay for the fees only.

Reply:
@erikzhang But `Sender` will have to sign the transaction, right? So far we cannot add a scope to sender's witness. How about change sender type to `Cosigner`, too?

Reply:
`Sender` and `Cosigner` can be the same value.

Reply:
Thats a very.nice comment @MorganDream, I suppose default behavior will be EntryOnly on sender (unless overwritten by a cosigner). We should test that.

Reply:
we can rename Sender to Payer, or at least add a comment

Reply:
> I suppose default behavior will be CalledByEntry on sender (unless overwritten by a cosigner of same address)
@igormcoelho Thanks for your response.
I am afraid that the allowed duplication between `Sender` and `Cosigner` may delegate the intuitive naming of these 2 props. How can someone be `Sender` and `Cosigner` at the same time? This design would probably induct user to ignore `Sender`'s signing safety as there is no scope settings on `Sender` but `Sender` will sign the transaction anyway.

Reply:
> Maybe the perfect solution is to add scope setting in Witness. Have you considered about that?

Yes... this was the original design in fact. But after many discussion, we saw that it wasn't sole attribution of a Witness itself, but the control system behind it, that applies only to Transaction (IVerifiable) payload. Other payloads (such as blocks) don't need this management, since it only applies to to Transaction Script running on Application mode.

This to emphasize: `CalledByEntry` is the safest mode of all, that's why I'm suggesting it. Another option is, like I said, to embed tx payment system on System Trigger (not Verification or Application), which is more complicated, but would give an even stronger guarantee (not even Entry Script would be able to use it).

I hope we find a good naming solution to it.. I love this feature ;)

Reply:
Thanks for your detailed explanation, it's a brilliant solution.
And I love this feature, too!

Reply:
@MorganDream I thought again, and CalledByEntry will be enough, impossible to break its security.
Personally, I would rename cosigners to AttachedWitnesses... but do you really think it will be confusing to have cosigner overwritting sender?
For wallet creators, its easy: every transfer should create a cosigner, even if its GAS one (redundand, yet good practice), so network fee can be freely used for its only purpose: attaching gas.
I used to.have another option for sender, which is send network fee as regular transfer to CN shared wallet. But this implies much more operations on every tx, just to.remove a field...
honestly, I dont.know how to.improve this.

Reply:
@igormcoelho 
> Personally, I would rename cosigners to AttachedWitnesses... but do you really think it will be confusing to have cosigner overwritting sender?

Yes. But not critical. User may be confused by why sender and cosigner can be the same one.
`AttachedWitnesses` is not so good, because may be regard as some appendix to `Witnesses` while it's not.
I don't catch you on the 2 cases you give. Anyway my proposal would be pushing sender and cosigners to one prop `Signers`, and add a Boolean prop `isFeePayer` to Cosigner Class.

Reply:
I get your point,your design is good, specially.if we want people to split fees... we embed network fee on Signer, remove the bool, since nonzero value is already payer. In naming,I would prefer this as a Senders field (not signer or cosigner),as signature may not be involved (it could be a password lock,timelock, or dapp verification..). Looks good to me.. lets open another issue?